parser/src/ast.rs

```diff
@@ -0,0 +1,86 @@
+use crate::lexer::Token;
+use std::fmt;
+
+/// [Identifier] is used to represent variable names.
+#[derive(Debug, Eq, PartialEq)]
+pub struct Identifier {
+    pub name: String,
+}
+
+impl fmt::Display for Identifier {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", &self.name)
+    }
+}
+
+/// [Source] is the root node of the AST representing the AIR constraints file.
+#[derive(Debug, PartialEq)]
+pub struct Source(pub Vec<SourceSection>);
+
+/// Source is divided into SourceSections.
+/// There are 4 types of Source Sections:
+/// - AirDef: Name of the air constraints module.
+/// - TraceCols: Trace Columns representing columns of the execution trace.
+/// - BoundaryConstraints: Boundary Constraints to be enforced on the boundaries of columns defined
+///   in the TraceCols section. Currently there are two types of boundaries, First and Last
+///   representing the first and last rows of the column.
+/// - TransitionConstraints: Transition Constraints to be enforced on the trace columns defined
+///   in the TraceCols section.
+#[derive(Debug, PartialEq)]
+pub enum SourceSection {
+    AirDef(AirDef),
+    TraceCols(TraceCols),
+    BoundaryConstraints(BoundaryConstraints),
+    TransitionConstraints(TransitionConstraints),
+}
+
+/// Name of the air constraints module.
+#[derive(Debug, Eq, PartialEq)]
+pub struct AirDef {
+    pub name: Identifier,
+}
+
+/// [TraceCols] contains the trace columns representing columns of the execution trace.
+/// TraceCols is made up of TraceColsGrp which represents groups of columns of a specific type.
+#[derive(Debug, PartialEq)]
+pub struct TraceCols {
+    pub main_cols: Vec<Identifier>,
+    pub aux_cols: Vec<Identifier>,
+}
+
+/// Stores the boundary constraints to be enforced to the trace column values.
+#[derive(Debug, PartialEq)]
+pub struct BoundaryConstraints {
+    pub boundary_constraints: Vec<Constraint>,
+}
+
+/// Describes the type of boundary in the boundary constraint.
+#[derive(Debug, Eq, PartialEq)]
+pub enum Boundary {
+    First,
+    Last,
+}
+
+/// Stores the transition constraints to be enforced to the trace column values.
+#[derive(Debug, PartialEq)]
+pub struct TransitionConstraints {
+    pub transition_constraints: Vec<Constraint>,
+}
+
+/// Stores the expression corresponding to the transition or boundary constraint.
+#[derive(Debug, PartialEq)]
+pub struct Constraint {
+    pub lhs: Expr,
+    pub rhs: Expr,
+}
+
+/// Arithmetic expressions representing transition or boundary constraint.
+#[derive(Debug, PartialEq)]
+pub enum Expr {
+    Add(Box<Expr>, Box<Expr>),
+    Subtract(Box<Expr>, Box<Expr>),
+    Equal(Box<Expr>, Box<Expr>),
```

Comment:
Do we need `Equal` to be a part of the expression? I can think of only a couple of places where we'd use it:

1. Constraint expression - which should already be covered.
2. Variable assignment - which we'll probably cover in a similar manner as the constraint expression.

Reply:
I was keeping it for variable assignment but makes sense to handle it the same way as constraint expression, so I've removed it. Thanks.